### ISSUE TITLE
[Attributor] Do not derive norecurse without looking at all call sites

### ISSUE THREAD

**Issue by jdoerfert** [llvm/llvm-project#53884]
Since we did not track the fact that we might have skipped call sites assumed dead in `checkForAllCallSites` we accidentally derive `norecurse` in certain situations.
An example (https://godbolt.org/z/oxvcjsaGf):

```
declare void @extern()

define i1 @test_rec_neg(i1 %c) norecurse {
    %rc1 = call i1 @rec(i1 true)
    br i1 %rc1, label %t, label %f
t:
    %rc2 = call i1 @rec(i1 %c)
    ret i1 %rc2
f:
    ret i1 %rc1
}

define internal i1 @rec(i1 %c1) {
    br i1 %c1, label %t, label %f
t:
    ret i1 true
f:
    %r = call i1 @rec(i1 true)
    call void @extern()
    ret i1 false
}
```

**Comment by jdoerfert** [comment]
/cherry-pick https://github.com/llvm/llvm-project/commit/6ed1ef0643aa900c283a6f0428c72a0d387e5ccc

**Comment by tstellar** [comment]
@jdoerfert Can you rebase the issue53884 branch? The patches no longer apply.

**Comment by tstellar** [comment]
Merged: f58ab32850211621d5986da2d687be0d550e7140
